api/src/main/java/bio/terra/tanagra/indexing/job/WriteAncestorDescendantIdPairs.java

```diff
@@ -0,0 +1,193 @@
+    super(entity);
+    this.hierarchyName = hierarchyName;
+  }
+
+  @Override
+  public String getName() {
+    return "WRITE ANCESTOR-DESCENDANT ID PAIRS ("
+        + getEntity().getName()
+        + ", "
+        + hierarchyName
+        + ")";
+  }
+
+  @Override
+  public void dryRun() {
+    run(true);
+  }
+
+  @Override
+  public void run() {
+    run(false);
+  }
+
+  private void run(boolean isDryRun) {
+    SQLExpression selectChildParentIdPairs =
+        getEntity()
+            .getSourceDataMapping()
+            .getHierarchyMapping(hierarchyName)
+            .queryChildParentPairs("child", "parent");
+    String sql = selectChildParentIdPairs.renderSQL();
+    LOGGER.info("select all ancestor-descendant id pairs SQL: {}", sql);
+
+    TablePointer outputTable =
+        getEntity()
+            .getIndexDataMapping()
+            .getHierarchyMapping(hierarchyName)
+            .getAncestorDescendant()
+            .getTablePointer();
+    BigQueryDataset outputBQDataset = getOutputDataPointer();
+    LOGGER.info(
+        "output BQ table: project={}, dataset={}, table={}",
+        outputBQDataset.getProjectId(),
+        outputBQDataset.getDatasetId(),
+        outputTable.getTableName());
+
+    String serviceAccountEmail;
+    try {
+      GoogleCredentials appDefaultSACredentials = GoogleCredentials.getApplicationDefault();
+      serviceAccountEmail = ((ServiceAccountCredentials) appDefaultSACredentials).getClientEmail();
+      LOGGER.info("Service account email: {}", serviceAccountEmail);
+    } catch (IOException ioEx) {
+      throw new SystemException("Error reading application default credentials.", ioEx);
+    }
+
+    // TODO: Allow overriding the default region.
+    String[] args = {
+      "--runner=dataflow",
+      "--project=" + outputBQDataset.getProjectId(),
+      "--region=" + DEFAULT_REGION,
+      "--serviceAccount=" + serviceAccountEmail,
+    };
+    // TODO: Use PipelineOptionsFactory.create() instead of fromArgs().
+    // PipelineOptionsFactory.create() doesn't seem to call the default factory classes, while
+    // fromArgs() does.
+    BigQueryOptions options =
+        PipelineOptionsFactory.fromArgs(args).withValidation().as(BigQueryOptions.class);
+
+    Pipeline pipeline = Pipeline.create(options);
+    PCollection<KV<Long, Long>> relationships =
+        pipeline
+            .apply(
+                "read parent-child query result",
+                BigQueryIO.readTableRows()
+                    .fromQuery(sql)
+                    .withMethod(BigQueryIO.TypedRead.Method.EXPORT)
+                    .usingStandardSql())
+            .apply(
+                MapElements.into(
+                        TypeDescriptors.kvs(TypeDescriptors.longs(), TypeDescriptors.longs()))
+                    .via(WriteAncestorDescendantIdPairs::relationshipRowToKV));
+    // TODO: Allow overriding the default max hierarchy depth.
```

Comment:
just curious, are the TODO tracked by tasks somewhere, or you will be tackling them in later PRs?

Reply:
There's no task tracking yet, because it's such a small team (backend is just me). I'm planning to tackle them in later PRs, but not urgently. We may start using JIRA once we onboard more engineers -- in that case, I'll start noting the ticket number in the comment as well.